examples/blake2f.rs

```diff
@@ -0,0 +1,1371 @@
+        BITS_NUMBER + 1,
+    ));
+    v1_vec_values[a] = (v1_vec_values[a] as u128 + v1_vec_values[b] as u128 + x as u128) as u64;
+
+    vd_bit_vec.push(split_value_4bits(v1_vec_values[d] as u128, BITS_NUMBER));
+    v1_vec_values[d] = ((v1_vec_values[d] ^ v1_vec_values[a]) >> R1)
+        ^ (v1_vec_values[d] ^ v1_vec_values[a]) << (64 - R1);
+    v_xor_d_bit_vec.push(split_value_4bits(v1_vec_values[d] as u128, BITS_NUMBER));
+
+    vc_bit_vec.push(split_value_4bits(
+        v1_vec_values[c] as u128 + v1_vec_values[d] as u128,
+        BITS_NUMBER + 1,
+    ));
+    v1_vec_values[c] = (v1_vec_values[c] as u128 + v1_vec_values[d] as u128) as u64;
+
+    vb_bit_vec.push(split_value_4bits(v1_vec_values[b] as u128, BITS_NUMBER));
+    v1_vec_values[b] = ((v1_vec_values[b] ^ v1_vec_values[c]) >> R2)
+        ^ (v1_vec_values[b] ^ v1_vec_values[c]) << (64 - R2);
+    v_xor_b_bit_vec.push(split_value_4bits(v1_vec_values[b] as u128, BITS_NUMBER));
+
+    va_bit_vec.push(split_value_4bits(
+        v1_vec_values[a] as u128 + v1_vec_values[b] as u128 + y as u128,
+        BITS_NUMBER + 1,
+    ));
+    v2_vec_values[a] = (v1_vec_values[a] as u128 + v1_vec_values[b] as u128 + y as u128) as u64;
+
+    vd_bit_vec.push(split_value_4bits(v1_vec_values[d] as u128, BITS_NUMBER));
+    v2_vec_values[d] = ((v1_vec_values[d] ^ v2_vec_values[a]) >> R3)
+        ^ (v1_vec_values[d] ^ v2_vec_values[a]) << (64 - R3);
+    v_xor_d_bit_vec.push(split_value_4bits(v2_vec_values[d] as u128, BITS_NUMBER));
+
+    vc_bit_vec.push(split_value_4bits(
+        v1_vec_values[c] as u128 + v2_vec_values[d] as u128,
+        BITS_NUMBER + 1,
+    ));
+    v2_vec_values[c] = (v1_vec_values[c] as u128 + v2_vec_values[d] as u128) as u64;
+
+    vb_bit_vec.push(split_value_4bits(v1_vec_values[b] as u128, BITS_NUMBER));
+    v2_vec_values[b] = ((v1_vec_values[b] ^ v2_vec_values[c]) >> R4)
+        ^ (v1_vec_values[b] ^ v2_vec_values[c]) << (64 - R4);
+    v_xor_b_bit_vec.push(split_value_4bits(
+        (v1_vec_values[b] ^ v2_vec_values[c]) as u128,
+        BITS_NUMBER,
+    ));
+    let bits = (v1_vec_values[b] ^ v2_vec_values[c]) / 2u64.pow(60);
+    b_bit_vec.push(F::from(bits / 8));
+    b_3bits_vec.push(F::from(bits % 8))
+}
+
+fn blake2f_circuit<F: PrimeField + Hash>(
+    ctx: &mut CircuitContext<F, InputValues>,
+    params: CircuitParams,
+) {
+    let v_vec: Vec<Queriable<F>> = (0..V_LEN)
+        .map(|i| ctx.forward(format!("v_vec[{}]", i).as_str()))
+        .collect();
+    let h_vec: Vec<Queriable<F>> = (0..H_LEN)
+        .map(|i| ctx.forward(format!("h_vec[{}]", i).as_str()))
+        .collect();
+    let m_vec: Vec<Queriable<F>> = (0..M_LEN)
+        .map(|i| ctx.forward(format!("m_vec[{}]", i).as_str()))
+        .collect();
+    let round = ctx.forward("round");
+
+    let blake2f_pre_step = ctx.step_type_def("blake2f_pre_step", |ctx| {
+        let v_vec = v_vec.clone();
+        let setup_v_vec = v_vec.clone();
```

Comment:
@10to4 setup_v_vec seems to be used in wg, perhaps should be called `wg_v_vec?

Cannot you do the clone inside wg, so it is more clear?

Reply:
Yes, that's right!

Reply:
The reason we clone vectors is that we can't use a same vector in both the setup function and the wg function, so we may be able to do the clone in wg, because that means we need to use v_vec in both the setup function and the wg function

Reply:
yes, I understand why we need to clone, but it is confusing that `setup_` is used in wg instead of setup. also if you can clone in the same level of where it is used, that is more clear.